connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -1156,12 +1161,12 @@ void fenceZombieSourceTasks(final ConnectorTaskId id, Callback<Void> callback) {
         fenceZombieSourceTasks(id.connector(), (error, ignored) -> {
             if (error == null) {
                 callback.onCompletion(null, null);
-            } else if (error instanceof NotLeaderException) {
+            } else if (error instanceof NotLeaderException && restClient != null) {
```

Comment:
Should we add a case for the `NotLeaderException` when `restClient == null`? We can still fail the request, but maybe add some more helpful information about why the request failed and possibly a hint about dedicated MM2 instances (which currently do not support intra-cluster communication).